service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java

```diff
@@ -0,0 +1,110 @@
+
+@Api
+@RestController
+@RequestMapping("/callback/placement")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class PlacementController {
+
+    private final ObjectMapper mapper;
+    private final PlacementService placementService;
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        boolean singleChild = placementService.hasSingleChild(caseData);
+
+        caseDetails.getData().put("singleChild", YesNo.from(singleChild));
+
+        if (singleChild) {
+            Element<Child> child = caseData.getAllChildren().get(0);
+            caseDetails.getData().put("placement", placementService.getPlacement(caseData, child));
+            caseDetails.getData().put("placementChildName", child.getValue().getParty().getFullName());
+        } else {
+            caseDetails.getData().put("childrenList", placementService.getChildrenList(caseData, null));
+        }
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        UUID childId = getSelectedChildId(caseDetails, caseData);
+        Element<Child> child = placementService.getChild(caseData, childId);
+
+        caseDetails.getData().put("childrenList", placementService.getChildrenList(caseData, child));
+        caseDetails.getData().put("placement", placementService.getPlacement(caseData, child));
+        caseDetails.getData().put("placementChildName", child.getValue().getParty().getFullName());
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSave(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        UUID childId = getSelectedChildId(caseDetails, caseData);
+        Element<Child> child = placementService.getChild(caseData, childId);
+
+        Placement placement = mapper.convertValue(caseDetails.getData().get("placement"), Placement.class)
+            .setChild(child);
+
+        caseDetails.getData().put("placements", placementService.setPlacement(caseData, placement));
+        removeFields(caseDetails, "placement", "placementChildName", "singleChild");
```

Comment:
How about `caseDetails.getData().keySet().removeAll(Set.of(...))`? We wouldn't need new method.

Reply:
the benefit of new method is that it gives you context what is happening and why (renamed method a bit)